### PR TITLE
Feat/tetris

### PR DESCRIPTION
**Pull request by kfwandhisfriends**

### CODE REVIEW

**Comment by PaulSeitz** [comment]
I went over everything again and I agree that all my initial concerns were adressed. I did not find anything else that needs fixing.
Your implementation looks very nice and after quite some time spent playing I did not encounter any bugs. 
 
Very well done!

**Comment by kfwandhisfriends** [comment]
> FYI you encountered an errenous test in your current test run (the first error with the first and last frame). You can ignore that error, we have removed the test until it is fixed :D
> 
> And for the other errors, they will be fixed by adding an 'all_rewards' attribute to your TetrisInfo(NamedTuple), as for example jax_pong has it. This can be used to enable users to provide different reward functions (also see jax_pong how it is set), I think you could just copy that.

thanks for your help again! 
we are implementing the scores all the way on top with LINE_CLEAR_SCORE = (0, 1, 2, 3, 4), and in do_soft_or_gravity method we give the player the choice, to add give points when blocks fall down. In our last version we have added all_rewards in it but also in other parts, in wich we had to add all_rewards, they are implemented with jnp.zeros(1). So the points are still determined at the top with other constants. Would it be possible, if you could take a look at our code, especially to the all_rewards parts and tell us if we have done it right?

**Comment by captainlmao774** [comment]
Hi, we are really happy to see that our game is merged, thank you so much! We wanted to add more comments to our code and ask you some questions before the deadline, but we couldn’t spend much time on it over the past two weeks because of exams. Still, we’d like to hear your opinion on a couple of things, if possible:)  

- We didn’t use step_counter like other example games in the _get_info function, instead return gives the state of the game, if a line is cleared and score with the function. Is it alright like this, or does it have to be done via step_counter like in the other games? 
- We include in our code extra functions to gain a score like, when falling down and when pressed space while it is set to hard drop. Which are not on by default and has to be set via changes in code. We thought of modern versions of tetris and left these just in case.
- The original game uses space to rotate blocks, which we also do but our implementation also uses up-key for block rotations. We initially thought it was convenient, so blocks can be rotated via space or up-keys. We wanted to ask, if it is a problem for testing, that the game is implemented with one extra key.   
  


  We would really appreciate your opinion on these. Please also let us know if you find anything that could be improved in our code. Thank you for your review!

**Comment by PaulSeitz** [comment]
> Hi, we are really happy to see that our game is merged, thank you so much! We wanted to add more comments to our code and ask you some questions before the deadline, but we couldn’t spend much time on it over the past two weeks because of exams. Still, we’d like to hear your opinion on a couple of things, if possible:)  
> 
> * We didn’t use step_counter like other example games in the _get_info function, instead return gives the state of the game, if a line is cleared and score with the function. Is it alright like this, or does it have to be done via step_counter like in the other games?
> * We include in our code extra functions to gain a score like, when falling down and when pressed space while it is set to hard drop. Which are not on by default and has to be set via changes in code. We thought of modern versions of tetris and left these just in case.
> * The original game uses space to rotate blocks, which we also do but our implementation also uses up-key for block rotations. We initially thought it was convenient, so blocks can be rotated via space or up-keys. We wanted to ask, if it is a problem for testing, that the game is implemented with one extra key.
> 
>   We would really appreciate your opinion on these. Please also let us know if you find anything that could be improved in our code. Thank you for your review!

Hey. Since we merged we game you can generally assume that we were very happy with the state of the implementation :)
Concerning your questions:
- the _get_info functions are in general rather flexible. Its not always necessary to have the step_counter in there, especially since potential users of this environment can modify the _get_info function to include whatever they need (this adaptability is actually a big advantage of these reimplementations in general)
- this is fine. Since they are deactivated by default it is something that is nice to have
- no thats fine, its a similar situation to the question about the _get_info function. Its something that can be adjusted if needed.

If you want to add documentation you can open a new pull request of course but as I said we are already very happy with your implementation, it is nicely done. Good job :)
